File: src/synapt_eval/suggestion_engine/rules/trending.py
"""Trending suggestion rules: multi-run pattern detection."""

from __future__ import annotations

from typing import Any

from synapt_eval.reviewer.types import SEVERITY_ERROR, SEVERITY_WARNING, Verdict
from synapt_eval.suggestion_engine.types import Suggestion
from synapt_eval.types import EvalResult


class MonotonicDegradationRule:
    """Detects 3+ consecutive runs where a key metric gets worse.

    Consumes run history from context["history"] (list of EvalResult,
    oldest first). Current result is included as the latest data point.
    """

    def __init__(
        self,
        metric: str = "p_at_5",
        consecutive: int = 3,
    ) -> None:
        self._metric = metric
        self._consecutive = consecutive

    def evaluate(
        self,
        result: EvalResult,
        verdicts: list[Verdict] | None = None,
        context: dict[str, Any] | None = None,
    ) -> list[Suggestion]:
        if not context or "history" not in context:
            return []

        history: list[EvalResult] = context["history"]
        same_category = [r for r in history if r.category == result.category]
        same_category.append(result)

        if len(same_category) < self._consecutive:
            return []

        values = [getattr(r.metrics, self._metric, None) for r in same_category]
        values = [v for v in values if v is not None]

        if len(values) < self._consecutive:
            return []

        tail = values[-self._consecutive :]
        monotonic_down = all(tail[i] > tail[i + 1] for i in range(len(tail) - 1))

        if not monotonic_down:
            return []

        return [
            Suggestion(
                severity=SEVERITY_ERROR,
                message=(
                    f"{self._metric} has degraded for {self._consecutive} consecutive runs "
                    f"in category '{result.category}': "
                    f"{' -> '.join(f'{v:.3f}' for v in tail)}"
                ),
                rule_name="monotonic_degradation",
                metric=self._metric,
                category=result.category,
                fix_hint=(
                    "Investigate recent changes; this metric is trending consistently downward."
                ),
            )
        ]


class StableLowRule:
    """Detects metrics stuck below threshold across multiple runs.

    Flags when a metric stays below the threshold for N or more
    consecutive runs without improvement, suggesting systemic issues.
    """

    def __init__(
        self,
        metric: str = "p_at_5",
        threshold: float = 0.7,
        min_runs: int = 3,
    ) -> None:
        self._metric = metric
        self._threshold = threshold
        self._min_runs = min_runs

    def evaluate(
        self,
        result: EvalResult,
        verdicts: list[Verdict] | None = None,
        context: dict[str, Any] | None = None,
    ) -> list[Suggestion]:
        if not context or "history" not in context:
            return []

        history: list[EvalResult] = context["history"]
        same_category = [r for r in history if r.category == result.category]
        same_category.append(result)

        if len(same_category) < self._min_runs:
            return []

        values = [getattr(r.metrics, self._metric, None) for r in same_category]
        values = [v for v in values if v is not None]

        if len(values) < self._min_runs:
            return []

        tail = values[-self._min_runs :]
        all_below = all(v < self._threshold for v in tail)

        if not all_below:
            return []

        avg = sum(tail) / len(tail)
        return [
            Suggestion(
                severity=SEVERITY_WARNING,
                message=(
                    f"{self._metric} has been below {self._threshold:.2f} for "
                    f"{self._min_runs} consecutive runs in category '{result.category}' "
                    f"(avg: {avg:.3f})"
                ),
                rule_name="stable_low",
                metric=self._metric,
                category=result.category,
                fix_hint=(
                    "This metric is consistently underperforming; "
                    "consider architectural changes rather than incremental tuning."
                ),
            )
        ]

File: ts/src/suggestion_engine/rules/trending.ts
import { SEVERITY_ERROR, SEVERITY_WARNING } from "../../reviewer/types.ts";
import type { SuggestionRule } from "../protocol.ts";
import type { Suggestion } from "../types.ts";
import type { EvalResult } from "../../types.ts";

type MetricKey = "pAt5" | "rAt10" | "tau" | "n";

export class MonotonicDegradationRule implements SuggestionRule {
  private readonly metric: MetricKey;
  private readonly consecutive: number;

  constructor(metric: MetricKey = "pAt5", consecutive = 3) {
    this.metric = metric;
    this.consecutive = consecutive;
  }

  evaluate(
    result: EvalResult,
    _verdicts?: unknown,
    context?: Record<string, unknown>,
  ): Suggestion[] {
    if (!context || !context["history"]) return [];

    const history = context["history"] as EvalResult[];
    const sameCategory = [
      ...history.filter((r) => r.category === result.category),
      result,
    ];

    if (sameCategory.length < this.consecutive) return [];

    const values = sameCategory
      .map((r) => r.metrics[this.metric])
      .filter((v): v is number => v != null);

    if (values.length < this.consecutive) return [];

    const tail = values.slice(-this.consecutive);
    const monotonicDown = tail.every(
      (v, i) => i === 0 || tail[i - 1] > v,
    );

    if (!monotonicDown) return [];

    const metricName = this.metricDisplayName();
    return [
      {
        severity: SEVERITY_ERROR,
        message: `${metricName} has degraded for ${this.consecutive} consecutive runs in category '${result.category}': ${tail.map((v) => v.toFixed(3)).join(" -> ")}`,
        ruleName: "monotonic_degradation",
        metric: metricName,
        category: result.category,
        fixHint:
          "Investigate recent changes; this metric is trending consistently downward.",
      },
    ];
  }

  private metricDisplayName(): string {
    const map: Record<MetricKey, string> = {
      pAt5: "p_at_5",
      rAt10: "r_at_10",
      tau: "tau",
      n: "n",
    };
    return map[this.metric];
  }
}

export class StableLowRule implements SuggestionRule {
  private readonly metric: MetricKey;
  private readonly threshold: number;
  private readonly minRuns: number;

  constructor(
    metric: MetricKey = "pAt5",
    threshold = 0.7,
    minRuns = 3,
  ) {
    this.metric = metric;
    this.threshold = threshold;
    this.minRuns = minRuns;
  }

  evaluate(
    result: EvalResult,
    _verdicts?: unknown,
    context?: Record<string, unknown>,
  ): Suggestion[] {
    if (!context || !context["history"]) return [];

    const history = context["history"] as EvalResult[];
    const sameCategory = [
      ...history.filter((r) => r.category === result.category),
      result,
    ];

    if (sameCategory.length < this.minRuns) return [];

    const values = sameCategory
      .map((r) => r.metrics[this.metric])
      .filter((v): v is number => v != null);

    if (values.length < this.minRuns) return [];

    const tail = values.slice(-this.minRuns);
    const allBelow = tail.every((v) => v < this.threshold);

    if (!allBelow) return [];

    const avg = tail.reduce((s, v) => s + v, 0) / tail.length;
    const metricName = this.metricDisplayName();
    return [
      {
        severity: SEVERITY_WARNING,
        message: `${metricName} has been below ${this.threshold.toFixed(2)} for ${this.minRuns} consecutive runs in category '${result.category}' (avg: ${avg.toFixed(3)})`,
        ruleName: "stable_low",
        metric: metricName,
        category: result.category,
        fixHint:
          "This metric is consistently underperforming; consider architectural changes rather than incremental tuning.",
      },
    ];
  }

  private metricDisplayName(): string {
    const map: Record<MetricKey, string> = {
      pAt5: "p_at_5",
      rAt10: "r_at_10",
      tau: "tau",
      n: "n",
    };
    return map[this.metric];
  }
}
